### PR TITLE
Allow dependent fields in struct formats and types

### LINKED ISSUE #244: Dependent struct formats and struct types
This will allow dependent struct formats to have field dependencies, which correspond to how structs are parsed in a binary format. For example:

```
struct Points : Format {
    len : U32Be,
    data : FormatArray len { x : U32Be, y U32Be },
}
```

Note that `len` is bound as `len : Repr U32Be` in the typing context. 

Unlike DDC we aim to preserve as much dependency information in the host language as possible. To that end `Repr` will interpret dependent struct formats into dependent struct types (as opposed to a non-dependent structs). 

We need to be careful of maintaining dependencies when looking up fields in these records. For example:

```
struct MyArray {
    len : Int,
    data : Array len Point,
}

get_data (array : MyArray) =
    array.data; // type of body is `Array array.len Point`
```

This will require us to implement local environments. When doing so we'll also have to update the `Bool` and `Int` eliminators to capture their current environment (lest we introduce weird scoping behaviour!)

### PR DESCRIPTION
**Pull request by toothbrush**
Main branch renamed to ... `main`.  Reopening, but this is morally still #256 (there is relevant discussion there). 

This allows for dependencies between fields in struct formats and types. This allows us to express more interesting, data-dependent struct formats, for example:

```
struct MyData : Format {
    count : U32Be,
    data : FormatArray count U32Be,
}
```

Closes #244.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -3,11 +3,11 @@ name: ci
 on:
   push:
     branches:
-      - master
+      - main
```
Comment (toothbrush) [inline .github/workflows/ci.yml:6]: Oops, totally missed that! :facepalm:

File: fathom/src/lang/core.rs [hunk 2]
```diff
@@ -276,3 +280,145 @@ impl Default for Globals {
         Globals::new(entries)
     }
 }
+
+/// A [de Bruijn index][de-bruijn-index] in the local environment.
+///
+/// De bruijn indices describe an occurrence of a variable in terms of the
```
Comment (toothbrush) [inline fathom/src/lang/core.rs:286]:
```suggestion
/// De Bruijn indices describe an occurrence of a variable in terms of the
```

Nitpick!

File: fathom/src/lang/core.rs [hunk 3]
```diff
@@ -276,3 +280,145 @@ impl Default for Globals {
...
+/// This describes an occurrence of a variable by counting the binders from
+/// the top of the term towards until the occurrence is reached.
+/// For example:
+///
+/// | Representation    | Example (S combinator)  |
+/// | ----------------- | ----------------------- |
+/// | Named             | `λx. λy. λz. x z (y z)` |
+/// | De Bruijn levels  | `λ_. λ_. λ_. 0 2 (1 2)` |
+///
+/// Levels are used in [values][semantics::Value] because they are not context-
```
Comment (toothbrush) [inline fathom/src/lang/core.rs:325]: Interesting!  I was unaware of this technique :bulb:

**Comment by toothbrush** [comment]
So far i've mainly digested the test case fixtures, but that looks really good!

File: fathom/src/lang/core.rs (hunk 2 above)
Comment (brendanzab) [reply fathom/src/lang/core.rs:286]: Oooh thanks!

File: fathom/src/lang/core.rs (hunk 3 above)
Comment (brendanzab) [reply fathom/src/lang/core.rs:325]: Yeah it's a super nifty way of doing dependent type checking

Comment (wezm) [reply fathom/src/lang/core.rs:325]: > This describes an occurrence of a variable by counting the binders from the top of the term towards until the occurrence is reached.

Is there a word missing after towards?

Comment (brendanzab) [reply fathom/src/lang/core.rs:325]: Good catch! Rewording this to:

> This describes an occurrence of a variable by counting the binders inwards from the top of the term until the occurrence is reached.

**Pull request merged**